psutil/_psutil_bsd.c

```diff
@@ -391,6 +396,104 @@ psutil_proc_cmdline(PyObject *self, PyObject *args) {
 }
 
 
+/*
+ * Return process environment as a Python dictionary
+ */
+PyObject *
+psutil_proc_environ(PyObject *self, PyObject *args) {
+    unsigned int i, cnt = 0;
+    long pid;
+    char **envs;
+    char errbuf[_POSIX2_LINE_MAX];
+    kvm_t *kd;
+#ifdef PSUTIL_NETBSD
+    struct kinfo_proc2 *p;
+#else
+    struct kinfo_proc *p;
+#endif
+    PyObject *py_retdict;
+
+    if (!PyArg_ParseTuple(args, "l", &pid))
+        return NULL;
+
+#if defined(PSUTIL_FREEBSD)
+    kd = kvm_openfiles(NULL, "/dev/null", NULL, 0, errbuf);
+#else
+    kd = kvm_openfiles(NULL, NULL, NULL, KVM_NO_FILES, errbuf);
+#endif
+    if (!kd) {
+        PyErr_SetString(PyExc_RuntimeError, errbuf);
+        return NULL;
+    }
+
+#if defined(PSUTIL_FREEBSD)
+    p = kvm_getprocs(kd, KERN_PROC_PID, pid, (int *)&cnt);
+#elif defined(PSUTIL_OPENBSD)
+    p = kvm_getprocs(kd, KERN_PROC_PID, pid, sizeof(*p), &cnt);
+#elif defined(PSUTIL_NETBSD)
+    p = kvm_getproc2(kd, KERN_PROC_PID, pid, sizeof(*p), &cnt);
+#endif
+
+    if (!p)
+    {
+        PyObject *ret = NoSuchProcess(kvm_geterr(kd));
```

Comment:
I'm not completely sure this is the correct action. 
`kvm_getprocs` [doc](https://www.freebsd.org/cgi/man.cgi?query=kvm_getprocs&sektion=3&n=1) is not clear on whether you can/should use `kvm_geterr`, but I suppose you can. What I'm worried about here is that the error may not necessarily be a string like "no such process". Perhaps it can also be "access denied". Or something else (in which case we want `RuntimeError`). So you should probably parse the string and decide what to do.

You could try a test: call this function for a PID which does not exist. If `kvm_geterr` returns a string indicating that, then it means you can rely on `kvm_geterr`. 

In that case, what I recommend, is reusing and extending `convert_kvm_error` utility function (check if string is something like "no such process"), after you move it into `psutil/_psutil_common.c` (see my comment above).

The resulting code here should be something like this:

```c
    if (!p) {
        convert_kvm_err("kvm_getprocs", kvm_geterr(kd));
        kvm_close(kd);
        return NULL;
    }
```

Reply:
I've checked the  `libkvm` sources and `/usr/src/bin/ps/ps.c` on all three BSD variants; they all use `kvm_err()` if there's a failing `kvm_*()` call.

I've written a tiny standalone test program to check for all return values (on all three platforms).
Fortunately, they all behave the same.

- `kvm_open*` isn't likely to fail with `EACCESS`, as it doesn't open `/dev/[k]mem` as in the old days.
It will only fail if the library cannot allocate internal memory, or cannot open the kernel boot file. `errno` will be set appropriate.
The question is which exception should be raised - `RuntimeError` or `OSError`?

- `kvm_getprocs` isn't likely to fail _itself_.
If the search for the PID fails, `cnt` is set to `0`. If the function itself returns `NULL`, it means that it was not able to allocate memory for the result, or the call parameters were invalid due to a programming error (`&cnt` pointing to invalid storage).
We can raise `RuntimeError` or `OSError` if the call itself failed.
If the call reports `cnt == 0`, this maps to `ESRCH` ("no such process").

I understand your concern mapping system errors to different Python errors;
now from what I've tested, `errno` is set appropriate after all failing `libkvm` calls.

There is just _one_ race which comes to my mind: When calling `kvm_getprocs()`, it may return success.
But before our call to `kvm_getenvv()`, we might have been preempted by the kernel, and the process in question might have terminated.
In this case, our later call to `kvm_getenvv()` will return `EINVAL`, and we probably should remap this to `ESRCH`?

Reply:
> kvm_open* isn't likely to fail with EACCESS, as it doesn't open /dev/[k]mem as in the old days.
It will only fail if the library cannot allocate internal memory, or cannot open the kernel boot file. errno will be set appropriate.
The question is which exception should be raised - RuntimeError or OSError?

If `errno` is set then it's easy: just use `PyErr_SetFromErrno(PyExc_OSError)`. My understanding though is that `kvm_` functions do not set `errno`, so we need to parse `kvm_geterr`. As a general rule, if `errno` is not set, parse `kvm_geterr` result and:

- set `ESRCH` if the string indicates the process is gone (use `NoSuchProcess("...");`)
- set `EPERM` if the string indicates some lack of privileges (use `AccessDenied("...");`)
- raise/call `PyErr_NoMemory()` if the string indicates it could not allocate memory
- raise `RuntimeError` in all other circumstances; if you can, also include the result of `kvm_geterr`

This logic (3 `if`s, 1 `else`) should be handled by `convert_kvm_err` utility function, and (if you feel like it, while you're at it =)) reused by all functions that use `kvm_*` APIs.
Maybe `convert_kvm_err` could have another `if` condition, which is executed first:

```c
if (errno != 0)
       PyErr_SetFromErrno(PyExc_OSError)
````

...but this assumes that one of the previous `kvm_` calls **did** set `errno`, but again, my understanding is that they usually don't (the various man pages never mention `errno`).

> kvm_getprocs isn't likely to fail itself. If the search for the PID fails, cnt is set to 0 [...]

Then raise `NoSuchProcess("...");`

> [...] If the function itself returns NULL, it means that it was not able to allocate memory for the result, or the call parameters were invalid due to a programming error (&cnt pointing to invalid storage, kd invalid or NULL, meaning a previous kvm_open* did not suceed). We can raise RuntimeError or OSError if the call itself failed.

Mmmm... if we have a way to differentiate with `kvm_geterr` (is it ENOMEM? is it EINVAL? ....) then I would say to just stick with `convert_kvm_err` utility fun. 

> If the call reports cnt == 0, this maps to ESRCH ("no such process"). This will also happen if the process with the given PID exists, but sysctl security.bsd.see_other_uids == 0.

Damn... This I don't know. =)

> There is just one race which comes to my mind: When calling kvm_getprocs(), it may return success.
But before our call to kvm_getenvv(), we might have been preempted by the kernel, and the process in question might have terminated. In this case, our later call to kvm_getenvv() will return EINVAL, and we probably should remap this to ESRCH?

Mmm... hard call. I would say yes, let's map it to `ESRCH`.

Reply:
The latest update of the pull requests contains the changes I've mentioned so far.

Now `psutil/tests/test_process.py` passes all `Process().environ()` related stuff on all `*BSD` variants.

Unfortunately, CirrusCI is now not happy with these latest changes.
Interestingly, the version yesterday passed the checks, which makes me a bit puzzled.

```code
ERROR: psutil.tests.test_contracts.TestFetchAllProcesses.test_all

File "/tmp/cirrus-ci-build/psutil/_psbsd.py", line 674, in environ
    return cext.proc_environ(self.pid)
OSError: [Errno 22] Invalid argument
```

Is it possible that there is some incosistency what test_process.py checks, and what test_contracts.py checks? It looks that someone is expecting an `OSError` exception where the other one expects a `NoSuchProcess` exception?

The whole test suite does not pass altogether, though it looks that it doesn't pass on the `5.7.2` release tag either?
That happens on all `*BSD` platforms here..

Reply:
> My understanding though is that kvm_ functions do not set errno, so we need to parse kvm_geterr

Actually, almost _all_ library functions do some system calls, *especially* the `kvm*` stuff.
So, after _any_ system call which failed, there will be `errno` set.

Usually library utility functions do not clear `errno` if they return an error condition themself.

This avoids inventing a dozen more library error codes if most stuff may be mapped on typical `errno` codes anyway.

Reply:
> Actually, almost all library functions do some system calls, especially the kvm* stuff. So, after any system call which failed, there will be errno set.

Got it. Then I think it makes sense to do this:

```c
#if defined(PSUTIL_FREEBSD) || defined(PSUTIL_OPENBSD) || defined(PSUTIL_NETBSD)
void
convert_kvm_err(const char *syscall, char *errbuf) {
    char fullmsg[8192];

    sprintf(fullmsg, "(originated from %s: %s)", syscall, errbuf);
    if (errno != 0)
        PyErr_SetFromOSErrnoWithSyscall(syscall);
    else if (strstr(errbuf, "Permission denied") != NULL)
        AccessDenied(fullmsg);
    else if (strstr(errbuf, "Operation not permitted") != NULL)
        AccessDenied(fullmsg);
    else
        PyErr_Format(PyExc_RuntimeError, fullmsg);
}
#endif
```

As for `EINVAL`, it's not clear to me where it originates from in `psutil_proc_environ` function. You should put `printf()`s to figure that out. At that point I'd say you'll have to decide what to do: whether to raise `NoSuchProcess` or `AccessDenied`.
From C you can use `psutil_pid_exists` to check if the PID exists. If it does raise AD, else NSP. It's done elsewhere in the C code: grep for `psutil_pid_exists`.

Reply:
...e.g. here:
https://github.com/giampaolo/psutil/blob/5e47e0bcba7585909a11e9c7424e864fc26a7fd2/psutil/arch/osx/process_info.c#L166-L174
Even if this is OSX, it may be that on BSD `kvm_getenvv` may raise `EINVAL` in case of zombie process.